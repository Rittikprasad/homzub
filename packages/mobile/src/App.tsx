import 'react-native-gesture-handler';
import React from 'react';
import { Provider } from 'react-redux';
import NetInfo from '@react-native-community/netinfo';
import FlashMessage, { MessageComponentProps } from 'react-native-flash-message';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { AnalyticsService } from '@homzhub/common/src/services/Analytics/AnalyticsService';
import { LinkingService } from '@homzhub/mobile/src/services/LinkingService';
import { I18nService } from '@homzhub/common/src/services/Localization/i18nextService';
import { NotificationService } from '@homzhub/mobile/src/services/NotificationService';
import { StoreProviderService } from '@homzhub/common/src/services/StoreProviderService';
import { IUserTokens, StorageKeys, StorageService } from '@homzhub/common/src/services/storage/StorageService';
import { configureStore } from '@homzhub/common/src/modules/store';
import { CommonActions } from '@homzhub/common/src/modules/common/actions';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { RootNavigator } from '@homzhub/mobile/src/navigation/RootNavigator';
import ErrorBoundary from '@homzhub/mobile/src/components/HOC/ErrorBoundary';
import { Toast } from '@homzhub/common/src/components/molecules/Toast';
import AppUpdate from '@homzhub/mobile/src/components/organisms/AppUpdate';
import ReviewAndRefer from '@homzhub/mobile/src/components/organisms/ReviewAndRefer';
import { SupportedLanguages } from '@homzhub/common/src/services/Localization/constants';

interface IState {
  booting: boolean;
}

StoreProviderService.init(configureStore);
const store = StoreProviderService.getStore();

export default class App extends React.PureComponent<{}, IState> {
  public listener: any;
  public state = {
    booting: true,
  };

  public componentDidMount = async (): Promise<void> => {
    this.listener = NetInfo.addEventListener((state) => {
      this.handleConnectivityChange(state.isConnected);
    });
    await LinkingService.firebaseInit();
    await this.bootUp();
    await AnalyticsService.initMixpanel();
    await NotificationService.init();
  };

  public componentWillUnmount(): void {
    this.listener();
  }

  public render = (): React.ReactNode => {
    const { booting } = this.state;

    return (
      <ErrorBoundary>
        <Provider store={store}>
          <RootNavigator booting={booting} />
          <FlashMessage position="bottom" MessageComponent={this.renderToast} />
          <AppUpdate />
          <ReviewAndRefer />
        </Provider>
      </ErrorBoundary>
    );
  };

  private renderToast = (props: MessageComponentProps): React.ReactElement => <Toast {...props} />;

  private handleConnectivityChange = (isConnected: boolean): void => {
    if (!isConnected) {
      AlertHelper.error({
        message: I18nService.t('common:serverError'),
      });
    }
  };

  private bootUp = async (): Promise<void> => {
    // FETCH COUNTRY OF THE DEVICE

    const isOnBoardingCompleted = (await StorageService.get<boolean>(StorageKeys.IS_ONBOARDING_COMPLETED)) ?? false;
    store.dispatch(CommonActions.getCountries());
    store.dispatch(UserActions.updateOnBoarding(isOnBoardingCompleted));

    const userData = await StorageService.get<IUserTokens>(StorageKeys.USER);
    if (userData) {
      store.dispatch(UserActions.loginSuccess(userData));
    }

    const selectedLanguage: SupportedLanguages | null = await StorageService.get(StorageKeys.USER_SELECTED_LANGUAGE);
    await I18nService.init(selectedLanguage || undefined);

    this.setState({ booting: false });
  };
}
