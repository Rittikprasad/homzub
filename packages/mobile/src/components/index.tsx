// Atoms
export { DonutGraph } from '@homzhub/mobile/src/components/atoms/DonutGraph';
export { DoubleBarGraph } from '@homzhub/mobile/src/components/atoms/DoubleBarGraph';
export { MoreProfile } from '@homzhub/mobile/src/components/atoms/MoreProfile';
export { PaginationComponent } from '@homzhub/mobile/src/components/atoms/PaginationComponent';
export { SnapCarousel } from '@homzhub/mobile/src/components/atoms/Carousel';
export { RNVideo } from '@homzhub/mobile/src/components/atoms/Video';
export { YoutubeVideo } from '@homzhub/mobile/src/components/atoms/YoutubeVideo';

// Molecules
export { AssetAdvertisementBanner } from '@homzhub/mobile/src/components/molecules/AssetAdvertisementBanner';
export { AssetDetailsImageCarousel } from '@homzhub/mobile/src/components/molecules/AssetDetailsImageCarousel';
export { AssetMarketTrends } from '@homzhub/mobile/src/components/molecules/AssetMarketTrends';
export { AssetMetrics } from '@homzhub/mobile/src/components/molecules/AssetMetrics';
export { AssetSummary } from '@homzhub/mobile/src/components/molecules/AssetSummary';
export { BottomSheetListView } from '@homzhub/mobile/src/components/molecules/BottomSheetListView';
export { ButtonGroup } from '@homzhub/mobile/src/components/molecules/ButtonGroup';
export { CollapsibleSection } from '@homzhub/mobile/src/components/molecules/CollapsibleSection';
export { CurrentLocation } from '@homzhub/mobile/src/components/molecules/CurrentLocation';
export { DetailsCard } from '@homzhub/mobile/src/components/molecules/DetailsCard';
export { DocumentCard } from '@homzhub/mobile/src/components/molecules/DocumentCard';
export { FullScreenAssetDetailsCarousel } from '@homzhub/mobile/src/components/molecules/FullScreenAssetDetailsCarousel';
export { Header } from '@homzhub/mobile/src/components/molecules/Header';
export { HeaderCard } from '@homzhub/mobile/src/components/molecules/HeaderCard';
export { LeaseProgress } from '@homzhub/mobile/src/components/molecules/LeaseProgress';
export { PaymentGateway } from '@homzhub/mobile/src/components/molecules/PaymentGateway';
export { PropertyDetailsLocation } from '@homzhub/mobile/src/components/molecules/PropertyDetailsLocation';
export { PropertyMapCard } from '@homzhub/mobile/src/components/molecules/PropertyMapCard';
export { GoogleSearchBar } from '@homzhub/mobile/src/components/molecules/GoogleSearchBar';
export { SearchResults } from '@homzhub/mobile/src/components/molecules/SearchResults';
export { ShieldGroup } from '@homzhub/mobile/src/components/molecules/ShieldGroup';
export { UploadBoxComponent } from '@homzhub/mobile/src/components/molecules/UploadBoxComponent';

// Organisms
export { AssetMetricsList } from '@homzhub/mobile/src/components/organisms/AssetMetricsList';
export type { IMetricsData } from '@homzhub/mobile/src/components/organisms/AssetMetricsList';
export { FinanceOverview } from '@homzhub/mobile/src/components/organisms/FinanceOverview';
export { LoginForm } from '@homzhub/common/src/components/organisms/LoginForm';
export { PaymentSuccess } from '@homzhub/mobile/src/components/organisms/PaymentSuccess';
export { PropertyDuesCardContainer } from '@homzhub/mobile/src/components/organisms/PropertyDuesCardContainer';
export { AuthenticationGateways } from '@homzhub/mobile/src/components/organisms/AuthenticationGateways';

// HOC
export { AnimatedProfileHeader } from '@homzhub/mobile/src/components/HOC/AnimatedProfileHeader';
export { StateAwareComponent } from '@homzhub/mobile/src/components/HOC/StateAwareComponent';
