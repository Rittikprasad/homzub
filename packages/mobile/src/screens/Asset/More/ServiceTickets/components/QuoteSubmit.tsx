import React, { ReactElement } from 'react';
import { StyleSheet } from 'react-native';
import { useTranslation } from 'react-i18next';
import { cloneDeep } from 'lodash';
import { useDispatch, useSelector } from 'react-redux';
import DocumentPicker from 'react-native-document-picker';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { TicketActions } from '@homzhub/common/src/modules/tickets/actions';
import { TicketSelectors } from '@homzhub/common/src/modules/tickets/selectors';
import { theme } from '@homzhub/common/src/styles/theme';
import { CollapsibleSection } from '@homzhub/mobile/src/components/molecules/CollapsibleSection';
import SubmitQuoteForm from '@homzhub/common/src/components/organisms/ServiceTickets/SubmitQuoteForm';
import { ICollapseSection } from '@homzhub/common/src/constants/ServiceTickets';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';

interface IProps {
  renderScreen: (children: React.ReactElement) => React.ReactElement;
  handleGoBack: () => void;
  setLoading: (isLoading: boolean) => void;
}

const QuoteSubmit = (props: IProps): ReactElement => {
  const { renderScreen, handleGoBack, setLoading } = props;
  const dispatch = useDispatch();
  const attachments = useSelector(TicketSelectors.getQuoteAttachment);
  const quotes = useSelector(TicketSelectors.getQuotes);
  const { t } = useTranslation(LocaleConstants.namespacesKey.serviceTickets);

  // HANDLERS

  const onUploadDoc = (index: number, tabIndex: number): void => {
    const prevQuotes = cloneDeep(quotes);
    const prevAttachments = cloneDeep(attachments);
    DocumentPicker.pick({
      type: [DocumentPicker.types.pdf, DocumentPicker.types.images],
    })
      .then((doc) => {
        prevQuotes[tabIndex].data[index].document = doc;
        // @ts-ignore
        dispatch(TicketActions.setQuoteAttachment([...prevAttachments, doc]));
        dispatch(TicketActions.setQuotes(prevQuotes));
      })
      .catch((e) => {
        if (!DocumentPicker.isCancel(e)) {
          AlertHelper.error({ message: e.message });
        }
      });
  };

  // HANDLERS

  const renderSection = (data: ICollapseSection): ReactElement => {
    return (
      <CollapsibleSection title={t(data.title)} containerStyle={styles.cardContainer} titleStyle={styles.cardTitle}>
        {data.children}
      </CollapsibleSection>
    );
  };

  const renderScreenData = (): React.ReactElement => {
    return (
      <SubmitQuoteForm
        renderCollapsibleSection={renderSection}
        onUploadDoc={onUploadDoc}
        onSuccess={handleGoBack}
        setLoader={setLoading}
      />
    );
  };

  return renderScreen(renderScreenData());
};

export default QuoteSubmit;

const styles = StyleSheet.create({
  cardContainer: {
    backgroundColor: theme.colors.gray10,
    paddingHorizontal: 16,
    paddingVertical: 12,
  },
  cardTitle: {
    color: theme.colors.darkTint1,
  },
});
