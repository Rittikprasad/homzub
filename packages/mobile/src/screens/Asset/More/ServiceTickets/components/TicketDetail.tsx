import React from 'react';
import { PickerItemProps, StyleSheet, View } from 'react-native';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { WithTranslation, withTranslation } from 'react-i18next';
import { LinkingService } from '@homzhub/mobile/src/services/LinkingService';
import { TicketActions } from '@homzhub/common/src/modules/tickets/actions';
import { TicketSelectors } from '@homzhub/common/src/modules/tickets/selectors';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import { EmptyState } from '@homzhub/common/src/components/atoms/EmptyState';
import { ImagePlaceholder } from '@homzhub/common/src/components/atoms/ImagePlaceholder';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';
import { AssetDetailsImageCarousel } from '@homzhub/common/src/components/molecules/AssetDetailsImageCarousel';
import { BottomSheet } from '@homzhub/common/src/components/molecules/BottomSheet';
import { BottomSheetListView } from '@homzhub/mobile/src/components/molecules/BottomSheetListView';
import { FullScreenAssetDetailsCarousel } from '@homzhub/mobile/src/components/molecules/FullScreenAssetDetailsCarousel';
import TicketActivityCard from '@homzhub/common/src/components/molecules/TicketActivity';
import TicketDetailsCard from '@homzhub/common/src/components/molecules/TicketDetailsCard';
import { Attachment } from '@homzhub/common/src/domain/models/Attachment';
import { Ticket, TicketStatus } from '@homzhub/common/src/domain/models/Ticket';
import { TicketAction } from '@homzhub/common/src/domain/models/TicketAction';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { ICurrentTicket, ITicketState } from '@homzhub/common/src/modules/tickets/interface';
import { TicketActions as Actions } from '@homzhub/common/src/constants/ServiceTickets';

interface ITicketAction {
  title: string;
  onPress: () => void;
  isDisabled: boolean;
}

interface IDispatchProps {
  getTicketDetail: (payload: number) => void;
  clearState: () => void;
  sendTicketReminder: () => void;
}

interface IStateProps {
  currentTicket: ICurrentTicket | null;
  ticketDetails: Ticket | null;
  isLoading: boolean;
  ticketLoader: ITicketState['loaders'];
  ticketActions: TicketAction[];
}

interface IScreenState {
  activeSlide: number;
  isFullScreen: boolean;
  isActionSheet: boolean;
  selectedAction: string;
  hasClickedWorkDone: boolean;
  showRemindSheet: boolean;
}

interface IProps {
  renderScreen: (children: React.ReactElement) => React.ReactElement;
  handleNavigation: (action: string) => void;
  isFFMUser?: boolean;
}

type Props = WithTranslation & IDispatchProps & IStateProps & IProps;

class ServiceTicketDetails extends React.Component<Props, IScreenState> {
  public focusListener: any;

  public state = {
    activeSlide: 0,
    isFullScreen: false,
    isActionSheet: false,
    selectedAction: '',
    hasClickedWorkDone: false,
    showRemindSheet: false,
  };

  public render = (): React.ReactNode => {
    const { isActionSheet, selectedAction, showRemindSheet } = this.state;
    const { t, renderScreen } = this.props;

    const actionList = this.getActionList();

    return (
      <>
        {renderScreen(this.renderChildren())}
        {this.renderActionButton()}
        {this.renderFullscreenCarousel()}
        <BottomSheetListView
          selectedValue={selectedAction}
          listHeight={350}
          data={actionList}
          listTitle={t('chooseAction')}
          onSelectItem={this.handleAction}
          onCloseDropDown={(): void => this.handleActionSheet(false)}
          isBottomSheetVisible={isActionSheet}
          extraContent={showRemindSheet ? this.renderRemindSheet : undefined}
        />
      </>
    );
  };

  private renderChildren = (): React.ReactElement => {
    const { t, ticketDetails, isLoading } = this.props;

    return (
      <View style={styles.container}>
        {!isLoading && !ticketDetails && <EmptyState isIconRequired={false} title={t('serviceTickets:noDetail')} />}
        {this.renderDetailsCard()}
        {this.renderActivityCard()}
      </View>
    );
  };

  private renderDetailsCard = (): React.ReactElement | null => {
    const { ticketDetails, isFFMUser } = this.props;
    if (!ticketDetails) return null;
    return (
      <TicketDetailsCard
        ticketData={ticketDetails}
        isFFMUser={isFFMUser}
        ticketImages={this.renderCarousel(ticketDetails)}
        renderRatingForm={this.renderRatingSheet}
        successCallback={this.onSuccessCallback}
      />
    );
  };

  private renderCarousel = (detail: Ticket): React.ReactElement => {
    const { activeSlide, isFullScreen } = this.state;
    if (!detail.ticketAttachments.length) {
      return <ImagePlaceholder height="100%" containerStyle={styles.carousel} />;
    }

    return (
      <>
        {!isFullScreen && (
          <AssetDetailsImageCarousel
            enterFullScreen={this.enableFullScreenWithImage}
            data={detail.ticketAttachments}
            activeSlide={activeSlide}
            updateSlide={this.updateSlide}
            containerStyles={styles.carousel}
            hasOnlyImages
          />
        )}
      </>
    );
  };

  private renderFullscreenCarousel = (): React.ReactElement | null => {
    const { isFullScreen, activeSlide, hasClickedWorkDone } = this.state;
    const { ticketDetails } = this.props;
    if (!isFullScreen || !ticketDetails) return null;
    let workDoneAttachments: Attachment[] = [];
    if (ticketDetails.status === TicketStatus.CLOSED) {
      const { activities } = ticketDetails;
      const lastActivity = activities[activities.length - 1];
      workDoneAttachments = lastActivity.data?.attachments ?? [];
    }
    return (
      <FullScreenAssetDetailsCarousel
        onFullScreenToggle={this.enableFullScreenWithImage}
        activeSlide={activeSlide}
        data={hasClickedWorkDone ? workDoneAttachments : ticketDetails.ticketAttachments}
        updateSlide={this.updateSlide}
        hasOnlyImages
      />
    );
  };

  private renderActivityCard = (): React.ReactElement | null => {
    const { ticketDetails, ticketActions, isFFMUser } = this.props;
    if (!ticketDetails) return null;

    // Check close ticket action is allowed or not
    const isCloseAllowed = ticketActions.find((item) => item.code === Actions.CLOSE_TICKET)?.isAllowed;
    return (
      <TicketActivityCard
        ticketData={ticketDetails}
        isFFMUser={isFFMUser}
        isCloseAllowed={isCloseAllowed}
        onPressQuote={this.handleQuoteClick}
        onPressImage={this.onFullScreenToggleCompleted}
      />
    );
  };

  private renderActionButton = (): React.ReactElement | null => {
    const { ticketDetails } = this.props;
    if (!ticketDetails) return null;
    const buttonData = this.getActionData();

    if (!buttonData) return null;
    return (
      <View style={styles.buttonContainer}>
        <Button type="primary" title={buttonData.title} disabled={buttonData.isDisabled} onPress={buttonData.onPress} />
        <Button
          type="primary"
          iconSize={20}
          icon={icons.downArrow}
          iconColor={theme.colors.white}
          containerStyle={styles.iconButton}
          onPress={(): void => this.handleActionSheet(true)}
        />
      </View>
    );
  };

  private renderRemindSheet = (): React.ReactElement => {
    const { t, sendTicketReminder } = this.props;
    const { showRemindSheet } = this.state;

    const onConfirmSendReminder = (): void => {
      sendTicketReminder();
      this.hideActionAndRemindSheets();
    };

    return (
      <BottomSheet
        visible={showRemindSheet}
        sheetHeight={theme.viewport.height / 2.8}
        onCloseSheet={this.hideRemindSheet}
        headerTitle={t('serviceTickets:remindToPay')}
        headerStyles={styles.remindHeader}
      >
        <View style={styles.remindSheetContent}>
          <Text type="small" style={styles.remindConfirmationText}>
            {t('serviceTickets:sendReminderConfirmation')}
          </Text>

          <View style={styles.remindInfoContainer}>
            <Icon name={icons.roundFilled} size={5} color={theme.colors.darkTint3} />
            <Label type="regular" textType="regular" style={styles.remindInfoText}>
              {t('serviceTickets:onceIn12Hours')}
            </Label>
          </View>

          <View style={styles.remindButtonsContainer}>
            <Button
              type="secondaryOutline"
              title={t('cancel')}
              titleStyle={styles.remindCancelTitle}
              onPress={this.hideRemindSheet}
              containerStyle={styles.remindCancelButton}
            />
            <Button type="primary" title={t('serviceTickets:remindNow')} onPress={onConfirmSendReminder} />
          </View>
        </View>
      </BottomSheet>
    );
  };

  private renderRatingSheet = (children: React.ReactElement, onClose: () => void): React.ReactElement => {
    const { t } = this.props;
    return (
      <BottomSheet visible headerTitle={t('serviceTickets:ticketReview')} sheetHeight={600} onCloseSheet={onClose}>
        {children}
      </BottomSheet>
    );
  };

  // HANDLERS START

  private onFullScreenToggleCompleted = (slideNumber: number): void => {
    this.setState((prevState) => ({
      isFullScreen: !prevState.isFullScreen,
      activeSlide: slideNumber,
      hasClickedWorkDone: true,
    }));
  };

  private onSuccessCallback = (): void => {
    const { currentTicket, getTicketDetail } = this.props;
    if (currentTicket) {
      getTicketDetail(currentTicket.ticketId);
    }
  };

  private handleActionSheet = (isOpen: boolean): void => {
    this.setState({ isActionSheet: isOpen });
  };

  private enableFullScreenWithImage = (): void => {
    this.setState((prevState) => ({
      isFullScreen: !prevState.isFullScreen,
      hasClickedWorkDone: false,
    }));
  };

  private updateSlide = (slideNumber: number): void => {
    this.setState({ activeSlide: slideNumber });
  };

  private handleQuoteClick = async (url: string): Promise<void> => {
    await LinkingService.canOpenURL(url);
  };

  // Take action button logic
  private getActionData = (): ITicketAction | null => {
    const { ticketActions } = this.props;
    if (!ticketActions) return null;

    const nextAction = ticketActions.filter((item) => item.isNext && item.code !== Actions.CLOSE_TICKET)[0];
    if (!nextAction) return null;
    return {
      title: nextAction.label,
      onPress: (): void => this.handleAction(nextAction.code),
      isDisabled: !nextAction.isAllowed,
    };
  };

  // Take action button handler
  private handleAction = (code: string): void => {
    const { handleNavigation } = this.props;
    const { isActionSheet } = this.state;
    handleNavigation(code);

    if (isActionSheet) {
      this.handleActionSheet(false);
    }
  };

  // Take action dropdown data logic
  private getActionList = (): PickerItemProps[] => {
    const { ticketActions } = this.props;
    if (!ticketActions) return [];

    // Excluding close ticket option from list
    return ticketActions
      .filter((actions) => actions.code !== Actions.CLOSE_TICKET)
      .map((item) => {
        return {
          label: item.label,
          value: item.code,
        };
      });
  };

  private showRemindSheet = (): void => {
    this.setState({ showRemindSheet: true });
  };

  private hideRemindSheet = (): void => {
    this.setState({ showRemindSheet: false });
  };

  private hideActionAndRemindSheets = (): void => {
    this.setState({ showRemindSheet: false, isActionSheet: false });
  };

  // HANDLERS END
}

const mapStateToProps = (state: IState): IStateProps => {
  const { getCurrentTicket, getTicketDetail, getTicketDetailLoader, getTicketLoaders, getTicketActions } =
    TicketSelectors;
  return {
    currentTicket: getCurrentTicket(state),
    ticketDetails: getTicketDetail(state),
    isLoading: getTicketDetailLoader(state),
    ticketLoader: getTicketLoaders(state),
    ticketActions: getTicketActions(state),
  };
};

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { getTicketDetail, clearState, sendTicketReminder } = TicketActions;
  return bindActionCreators(
    {
      getTicketDetail,
      clearState,
      sendTicketReminder,
    },
    dispatch
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(withTranslation()(ServiceTicketDetails));

const styles = StyleSheet.create({
  container: {
    marginTop: 16,
  },
  carousel: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    width: theme.viewport.width - 64,
    borderRadius: 10,
    height: 150,
    marginHorizontal: 16,
  },
  buttonContainer: {
    flexDirection: 'row',
    padding: 20,
    backgroundColor: theme.colors.white,
  },
  iconButton: {
    flex: 0.2,
    marginLeft: 2,
  },
  remindHeader: {
    fontSize: 20,
  },
  remindSheetContent: {
    paddingHorizontal: 16,
    flex: 1,
  },
  remindConfirmationText: {
    color: theme.colors.darkTint3,
  },
  remindInfoContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginTop: 10,
  },
  remindInfoText: {
    color: theme.colors.darkTint3,
    marginStart: 5,
  },
  remindButtonsContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    marginTop: 16,
  },
  remindCancelButton: {
    borderWidth: 1,
    borderColor: theme.colors.primaryColor,
    marginEnd: 16,
  },
  remindCancelTitle: {
    color: theme.colors.primaryColor,
  },
});
