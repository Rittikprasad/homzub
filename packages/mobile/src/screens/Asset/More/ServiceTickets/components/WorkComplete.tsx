import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import ImagePicker, { Image as ImagePickerResponse } from 'react-native-image-crop-picker';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { TicketRepository } from '@homzhub/common/src/domain/repositories/TicketRepository';
import { AnalyticsService } from '@homzhub/common/src/services/Analytics/AnalyticsService';
import { AttachmentService } from '@homzhub/common/src/services/AttachmentService';
import { TicketActions } from '@homzhub/common/src/modules/tickets/actions';
import { TicketSelectors } from '@homzhub/common/src/modules/tickets/selectors';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import ProofOfCompletion from '@homzhub/common/src/components/molecules/ProofOfCompletion';
import { IUploadAttachmentResponse } from '@homzhub/common/src/components/organisms/AddRecordForm';
import { ICompleteTicketPayload, TicketAction } from '@homzhub/common/src/domain/repositories/interfaces';
import { EventType } from '@homzhub/common/src/services/Analytics/EventType';
import { IImageSource } from '@homzhub/common/src/services/AttachmentService/interfaces';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { AttachmentType } from '@homzhub/common/src/constants/AttachmentTypes';
import { TOTAL_IMAGES } from '@homzhub/common/src/constants/ServiceTickets';

interface IProps {
  renderScreen: (children: React.ReactElement) => React.ReactElement;
  setLoader: (isLoading: boolean) => void;
  onSuccess: () => void;
}

const WorkComplete = (props: IProps): React.ReactElement => {
  const { renderScreen, setLoader, onSuccess } = props;
  const dispatch = useDispatch();

  const [comment, setComment] = useState('');
  const attachments = useSelector(TicketSelectors.getProofAttachment);
  const selectedTicket = useSelector(TicketSelectors.getCurrentTicket);
  const { t } = useTranslation(LocaleConstants.namespacesKey.serviceTickets);

  // HANDLERS
  const handleImageUpload = async (): Promise<void> => {
    try {
      const response: ImagePickerResponse | ImagePickerResponse[] = await ImagePicker.openPicker({
        compressImageMaxWidth: 400,
        compressImageMaxHeight: 400,
        compressImageQuality: PlatformUtils.isAndroid() ? 1 : 0.8,
        includeBase64: true,
        maxFiles: !attachments.length ? TOTAL_IMAGES : attachments.length - TOTAL_IMAGES,
        multiple: true,
        mediaType: 'photo',
      });

      const images = response as ImagePickerResponse[];
      let attachment: IImageSource[] = [];
      images.forEach((item) => {
        attachment = [
          ...attachment,
          {
            filename: item.filename,
            path: item.path,
            mime: item.mime,
          },
        ];
      });
      dispatch(TicketActions.setAttachment(attachment));
    } catch (e) {
      if (e.code !== 'E_PICKER_CANCELLED') {
        AlertHelper.error({ message: e.message });
      }
    }
  };

  const onCommentChange = (value: string): void => {
    setComment(value);
  };

  const onWorkDone = async (): Promise<void> => {
    const formData = new FormData();
    let attachmentIds: number[] = [];

    if (attachments.length > 10) {
      AlertHelper.error({ message: t('maxProof') });
      return;
    }

    setLoader(true);
    if (attachments.length > 0) {
      attachments.forEach((image: any) => {
        // @ts-ignore
        formData.append('files[]', {
          // @ts-ignore
          name: PlatformUtils.isIOS() ? image.filename : image.path.substring(image.path.lastIndexOf('/') + 1),
          uri: image.path,
          type: image.mime,
        });
      });

      /* API call for attachment upload on s3 */
      const response = await AttachmentService.uploadImage(formData, AttachmentType.TICKET_DOCUMENTS);
      const { data, error } = response;
      if (data && data.length > 0) {
        attachmentIds = data.map((i: IUploadAttachmentResponse) => i.id);
      }

      if (error) {
        AlertHelper.error({ message: t('common:fileCorrupt') });
      }
    }
    try {
      if (selectedTicket) {
        const payload: ICompleteTicketPayload = {
          param: { ticketId: selectedTicket.ticketId },
          data: {
            action: TicketAction.COMPLETE_TICKET,
            payload: {
              ...(!!comment && { comment }),
              attachments: attachmentIds,
            },
          },
        };

        await TicketRepository.completeTicket(payload);
        setLoader(false);
        AlertHelper.success({ message: t('ticketComplete') });
        onSuccess();
        dispatch(TicketActions.clearState());
        dispatch(TicketActions.getTickets());

        /* Analytics for closed ticket */
        AnalyticsService.track(EventType.ClosedServiceTicket, {
          project_name: selectedTicket.propertyName ?? '',
          ticketId: selectedTicket.ticketId,
        });
      }
    } catch (e) {
      setLoader(false);
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };

  // HANDLERS

  const renderScreenData = (): React.ReactElement => {
    return (
      <>
        <ProofOfCompletion onUploadImage={handleImageUpload} onUpdateComment={onCommentChange} />
        <Button type="primary" title={t('common:done')} onPress={onWorkDone} />
      </>
    );
  };

  return renderScreen(renderScreenData());
};

export default WorkComplete;
