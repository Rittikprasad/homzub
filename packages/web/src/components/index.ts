// atoms
export { SearchField } from '@homzhub/web/src/components/atoms/SearchField';

// molecules
export { default as MultiCarousel } from '@homzhub/web/src/components/molecules/MultiCarousel';
export { NextPrevBtn } from '@homzhub/web/src/components/molecules/NextPrevBtn';
export { default as Navbar } from '@homzhub/web/src/components/molecules/Navbar';
export { PendingPropertiesCard } from '@homzhub/web/src/components/molecules/PendingPropertiesCard';
export { NavigationInfo } from '@homzhub/web/src/components/molecules/NavigationInfo';
export { default as BreadCrumbs } from '@homzhub/web/src/components/molecules/BreadCrumbs';

// HOC
export { Hoverable } from '@homzhub/web/src/components/hoc/Hoverable';
export { StickyHeader } from '@homzhub/web/src/components/hoc/StickyHeader';

// organisms
