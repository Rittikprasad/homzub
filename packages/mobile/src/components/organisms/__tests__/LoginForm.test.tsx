// @ts-nocheck
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { LoginForm } from '@homzhub/mobile/src/components';

jest.mock('@homzhub/common/src/services/storage/StorageService', () => 'StorageService');
jest.mock('@react-native-community/google-signin', () => {});

describe.skip('Test cases for LoginForm', () => {
  const testProps = (Props: any): any => ({
    ...Props,
    onLoginSuccess: jest.fn(),
  });

  let props: any;

  it('should render snapshot', () => {
    props = testProps({});
    const wrapper = shallow(<LoginForm {...props} t={(key: string): string => key} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render snapshot for dropdown selection', () => {
    props = testProps({});
    const wrapper = shallow(<LoginForm {...props} t={(key: string): string => key} />);
    const instance = wrapper.dive().instance();
    expect(instance).toMatchSnapshot();
  });
});
