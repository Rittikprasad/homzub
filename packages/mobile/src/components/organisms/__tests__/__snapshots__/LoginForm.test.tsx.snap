// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test cases for LoginForm should render snapshot 1`] = `
<Connect(LoginForm)
  i18n={Object {}}
  onLoginSuccess={[MockFunction]}
  t={[Function]}
  tReady={false}
/>
`;

exports[`Test cases for LoginForm should render snapshot for dropdown selection 1`] = `
LoginForm {
  "context": Object {},
  "handleSubmit": [Function],
  "loginEmailFormSchema": [Function],
  "loginPhoneFormSchema": [Function],
  "password": Object {
    "current": null,
  },
  "props": Object {
    "i18n": Object {},
    "onLoginSuccess": [MockFunction],
    "t": [Function],
    "tReady": false,
  },
  "refs": Object {},
  "renderLoginFields": [Function],
  "setState": [Function],
  "state": Object {
    "email": "",
    "isEmailFlow": false,
    "password": "",
    "phone": "",
    "phoneCode": "",
  },
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useDeferredValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useResponder": [Function],
        "useState": [Function],
        "useTransition": [Function],
      },
      "_element": <LoginForm
        i18n={Object {}}
        onLoginSuccess={[MockFunction]}
        t={[Function]}
        tReady={false}
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <KeyboardAvoidingView
        behavior="padding"
        enabled={true}
        keyboardVerticalOffset={0}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Formik
          initialValues={
            Object {
              "email": "",
              "isEmailFlow": false,
              "password": "",
              "phone": "",
              "phoneCode": "",
            }
          }
          onSubmit={[Function]}
          validate={[Function]}
        >
          [Function]
        </Formik>
      </KeyboardAvoidingView>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;
